pkg/identityserver/storetest/end_device_store.go

```diff
@@ -460,3 +462,81 @@ func (st *StoreTest) TestEndDeviceStorePagination(t *T) {
 		}
 	})
 }
+
+func (st *StoreTest) TestEndDeviceBatchUpdate(t *T) {
+	usr1 := st.population.NewUser()
+	app1 := st.population.NewApplication(usr1.GetOrganizationOrUserIdentifiers())
+
+	var all []*ttnpb.EndDevice
+	for i := 0; i < 3; i++ {
+		all = append(all, st.population.NewEndDevice(app1.GetIds()))
+	}
+
+	dev1 := all[0]
+	dev2 := all[1]
+	dev3 := all[2]
+
+	s, ok := st.PrepareDB(t).(interface {
+		Store
+		is.EndDeviceStore
+	})
+	defer st.DestroyDB(t, false)
+	defer s.Close()
+	if !ok {
+		t.Fatal("Store does not implement EndDeviceStore")
+	}
+
+	t.Run("BatchUpdateEndDeviceLastSeen", func(t *T) {
+		a, ctx := test.New(t)
+
+		validDevTime := time.Now()
+		dev1.LastSeenAt = ttnpb.ProtoTimePtr(validDevTime)
+		dev2.LastSeenAt = ttnpb.ProtoTimePtr(validDevTime)
+		dev3.LastSeenAt = ttnpb.ProtoTimePtr(validDevTime.Add(10 * time.Second))
+
+		batch := []*ttnpb.BatchUpdateEndDeviceLastSeenRequest_EndDeviceLastSeenUpdate{
+			{Ids: dev1.Ids, LastSeenAt: dev1.LastSeenAt},
+			{Ids: dev2.Ids, LastSeenAt: dev2.LastSeenAt},
+			{Ids: dev3.Ids, LastSeenAt: dev3.LastSeenAt},
+		}
+		err := s.BatchUpdateEndDeviceLastSeen(ctx, batch)
+		a.So(err, should.BeNil)
+		got, err := s.ListEndDevices(ctx, app1.GetIds(), []string{"last_seen_at"})
+		a.So(err, should.BeNil)
+		a.So(got, should.HaveLength, 3)
+
+		for _, dev := range got {
+			a.So(dev.LastSeenAt, should.NotBeNil)
+		}
```

Comment:
Nit:

```suggestion
		if a.So(err, should.BeNil) && a.So(got, should.NotBeNil) && a.So(got, should.HaveLength, 3) {
			for _, dev := range got {
				a.So(dev.LastSeenAt, should.NotBeNil)
			}
		}
```

That way, if something doesn't go as expected, we'll just see the first failure and won't get confused by panics caused by subsequent assertions.

---

And should we make sure that the correct timestamp is set for the correct device (just copy-paste from below)?